File: LC-Java/src/List/_141_hasCycle.java
package List;

import DataStructure.ListNode;

public class _141_hasCycle {

    public static void main(String[] args) {
        ListNode head = new ListNode(0);
        ListNode p1 = new ListNode(1);
        ListNode p2 = new ListNode(2);
        ListNode p3 = new ListNode(3);
        ListNode p4 = new ListNode(4);
        ListNode p5 = new ListNode(5);
        ListNode p6 = new ListNode(6);
        ListNode p7 = new ListNode(7);
        ListNode p8 = new ListNode(8);
        ListNode p9 = new ListNode(9);

        p1.next=p2;
        p2.next=p3;
        p3.next=p4;
        p4.next=p5;
        p5.next=p6;
        p6.next=p7;
        // 环
        // p7.next=p4;
        System.out.println(new _141_hasCycle().hasCycle(p1));
    }
    public boolean hasCycle(ListNode head) {

        if (head==null||head.next==null) {
            return false;
        }

        ListNode fast=head,slow=head;
        while (fast != null || fast.next != null) {
            fast=fast.next.next;
            slow=slow.next;
            if (slow ==fast) {
                return true;
            }
        }
        return  false;
    }
}

File: TopKIntervierQuestions/Round-1/141_hasCycle.go
package round1

func hasCycle(head *ListNode) bool {
	if head == nil || head.Next == nil {
		return false
	}
	fast, slow := head.Next, head
	for fast != slow {
		if fast == nil || fast.Next == nil {
			return false
		}
		fast = fast.Next.Next
		slow = slow.Next
	}
	return true
}

// use set (implmented by struct{})
func hasCycle2(head *ListNode) bool {
	dict := make(map[*ListNode]struct{})
	cur := head
	for cur != nil {
		if _, ok := dict[cur]; ok {
			return true
		}
		dict[cur] = struct{}{}
		cur = cur.Next
	}
	return false
}
